sql/catalyst/src/main/scala/org/apache/spark/sql/types/YearMonthIntervalType.scala

```diff
@@ -55,16 +65,28 @@ class YearMonthIntervalType private() extends AtomicType {
 
   private[spark] override def asNullable: YearMonthIntervalType = this
 
-  override def typeName: String = "interval year to month"
+  override val typeName: String = (start, end) match {
+    case (YEAR, MONTH) => "interval year to month"
+    case (YEAR, YEAR) => "interval year"
+    case (MONTH, MONTH) => "interval month"
+    case _ => throw new AnalysisException(
```

Comment:
`IllegalArgumentException`?

Reply:
Since it's user-facing, can we put it in `QueryCompilationError`?

Reply:
I followed the existing approach:
https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala#L47-L55